[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32658 [JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API**
* #32604 [JIT] pickle serialization for custom bound classes
* #32586 [JIT] Include custom_class.h in torch/script.h
* #32555 [JIT] Serialize attributes and types in ClassType serialization
* #32553 [JIT] Support for registering C++ lambdas as methods on custom C++ class

Differential Revision: [D19584701](https://our.internmc.facebook.com/intern/diff/D19584701)

Commit:
[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit da2548c3:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=da2548c3795d78f59bcaaafbff48ae4bb10b915d).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4350303) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=91429732) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=91429732))
<details>
<summary>
<code>Jan 27 12:30:41 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 27 12:30:41 Ran 1047 tests in 295.302s 
Jan 27 12:30:41  
Jan 27 12:30:41 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 27 12:30:41  
Jan 27 12:30:41 Generating XML reports... 
Jan 27 12:30:41 Traceback (most recent call last): 
Jan 27 12:30:41   File "test/run_test.py", line 456, in <module> 
Jan 27 12:30:41     main() 
Jan 27 12:30:41   File "test/run_test.py", line 449, in main 
Jan 27 12:30:41     raise RuntimeError(message) 
Jan 27 12:30:41 RuntimeError: test_autograd failed! 
Jan 27 12:30:41 + cleanup 
Jan 27 12:30:41 + retcode=1 
Jan 27 12:30:41 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

jamesr66a commented on torch/custom_class_detail.h:
```diff
@@ -51,6 +37,51 @@ struct pickle_factory<Get, Set, RetState(Self), NewInstance(ArgState)> {
+  }
+
+  R (CurrClass::*m)(Args...) const;
+};
```
It would be nice if these two specialization could be the same thing. (they're not right now because of `const`)

smessmer replied:
It's probably possible to unify them into one but I'm not sure if it's worth the effort. Ping me offline if you want to try.

smessmer commented on torch/custom_class_detail.h:
```diff
@@ -51,6 +37,51 @@ struct pickle_factory<Get, Set, RetState(Self), NewInstance(ArgState)> {
+  WrapMethod(R (CurrClass::*m)(Args...)) : m(std::move(m)) {}
+
+  R operator()(c10::intrusive_ptr<CurrClass> cur, Args... args) {
+    return c10::guts::invoke(m, *cur, args...);
```
- you probably want to std::forward the args
- cur->(*m)(...) should work here and is easier to understand than a call to invoke().

Commit:
Update on "[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API"


Differential Revision: [D19584701](https://our.internmc.facebook.com/intern/diff/D19584701)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API"


Differential Revision: [D19584701](https://our.internmc.facebook.com/intern/diff/D19584701)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API"


Differential Revision: [D19584701](https://our.internmc.facebook.com/intern/diff/D19584701)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API"


Differential Revision: [D19584701](https://our.internmc.facebook.com/intern/diff/D19584701)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API"


Differential Revision: [D19584701](https://our.internmc.facebook.com/intern/diff/D19584701)

[ghstack-poisoned]
